README
qmail-notify
Delayed delivery notification for qmail
Bruce Guenter <bruceg@em.ca>
Version 0.92
2001-08-08

This package contains a program to notify senders about email that has
been held in the qmail queue.

A mailing list has been set up to discuss this and other packages.
To subscribe, send an email to:
	bgware-subscribe@lists.em.ca
A mailing list archive is available at:
	http://lists.em.ca/?list=bgware

Development versions of qmail-notify are available via anonymous CVS.
Set your CVSROOT to ":pserver:cvs@bruce-guenter.dyndns.org:/CVS",
login with an empty password, and check out the "qmail-notify" module.

Requirements:

- qmail 1.01-1.03, with or without the big-todo patch

- cron

Installation:

- As root, run:
	make
	./installer

Configuration:

- The simple usage is to put qmail-notify into an hourly cron job.  For
  more details on the options, run "qmail-notify -h".

- To use an alternate response message, save the message to a file and
  specify the path to the file with the "-f" option.  The file must
  contain three instances of "%s".  The first is replaced with the
  server's hostname; the second is replaced with the notification delay
  (controlled by the "-t" option); the third is replaced with the
  message lifetime (controled by the qmail control/queuelifetime file).

This program is Copyright(C) 2000 Bruce Guenter, and may be copied
according to the GNU GENERAL PUBLIC LICENSE (GPL) Version 2 or a later
version.  A copy of this license is included with this package.  This
package comes with no warranty of any kind.


NEWS
-------------------------------------------------------------------------------
Changes in version 1.0

- Now supports sending multiple notifications for a single message by
  allowing multiple message ages to be specified with the "-t" option.

- Allow for some of the hard-coded paths to be modified.

- Fixed handling of sender addresses starting with a hyphen.

- Fixed treating missing info files as fatal errors.

- Allows for configuring the base qmail directory in $QMAILHOME.

- Switched to using bg-installer for installation.

- Updated for bglibs v2.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.93

- Added support for morercpthosts.cdb.

- Now defaults to not sending a copy of the nofication.

- Recoded to use iobuf libraries.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------
Changes in version 0.92

- Added support for MIME-encoded attachments.

- Added support for loading an external bounce response message file.

Development of this version has been sponsored by FutureQuest, Inc.
ossi@FutureQuest.net  http://www.FutureQuest.net/
-------------------------------------------------------------------------------


openread.c
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include <bglibs/msg.h>
#include <bglibs/str.h>

#include "qmail-notify.h"

int open_file(const char* prefix, const char* filename)
{
  static str fullname;
  if (!str_truncate(&fullname, 0)) oom();
  if (prefix != 0) {
    if (!str_copys(&fullname, prefix)) oom();
    if (!str_catc(&fullname, '/')) oom();
  }
  if (!str_cats(&fullname, filename)) oom();
  return open(fullname.s, O_RDONLY);
}

char* read_file(const char* prefix, const char* filename)
{
  struct stat statbuf;
  char* data;
  int fd = open_file(prefix, filename);
  if(fd == -1 || fstat(fd, &statbuf) == -1) {
    data = malloc(1);
    data[0] = 0;
  }
  else {
    data = malloc(statbuf.st_size+1);
    if (read(fd, data, statbuf.st_size) != statbuf.st_size)
      die3sys(111, "Could not read data from '", filename, "'");
    data[statbuf.st_size] = 0;
  }
  close(fd);
  return data;
}

char* read_line(const char* filename)
{
  char* data = read_file(0, filename);
  char* nl = strchr(data, '\n');
  if(nl)
    *nl = 0;
  return data;
}

long read_int(const char* filename, long dflt)
{
  char* data = read_line(filename);
  long result = dflt;
  if(*data) {
    char* ptr;
    result = strtol(data, &ptr, 10);
    if(*ptr)
      result = dflt;
  }
  free(data);
  return result;
}


qmail-notify-make.c
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include <bglibs/systime.h>
#include <bglibs/obuf.h>
#include <bglibs/misc.h>
#include <bglibs/msg.h>

#include "qmail-notify.h"

static const char* bounce_header = "{"
"From: <MAILER-DAEMON@}s{>\n"
"To: <}s{>\n"
"Subject: delayed delivery notice\n}";

static const char* mime_bounce_header = "{"
"MIME-Version: 1.0\n"
"Content-Type: multipart/mixed; boundary=\"}s{\"\n"
"\n"
"This is a multi-part message in MIME format.\n"
"(If you can see this message, your E-mail client is not MIME compatible.)\n"
"--}s{\n"
"Content-Type: text/plain; charset=us-ascii\n"
"Content-Transfer-Encoding: 7bit\n}";

static const char* bounce_body = "{"
"Your message has been received by }s{ but has been\n"
"undeliverable to the following recipients for at least }s{.\n"
"The mail system will continue to attempt to deliver your message\n"
"to these recipients for a total of }s{.  You do not need to\n"
"resend your message at this time.\n}";

static const char* recipients_prefix = "\nRecipient(s):\n";

static const char* message_seperator =
"\n"
"--- Below this line is a copy of the original message.\n"
"\n";

static const char* mime_message_seperator = "{"
"\n"
"The following attachment contains a copy of the original message.\n"
"\n"
"--}s{\n"
"Content-Type: message/rfc822\n"
"Content-Disposition: inline\n"
"\n}";

static const char* mime_message_end = "{\n--}s{--}";

static char mime_boundary[65];

static void make_mime_boundary(void)
{
  static char chars[] = "0123456789ABCDEFHIJKLMNOPQRTUVWX";
  int i;
  struct timeval now;
  gettimeofday(&now, 0);
  random_init(now.tv_sec ^ now.tv_usec ^ getpid());
  for(i = 0; i < 64; i++)
    mime_boundary[i] = chars[random_trunc(32)];
  mime_boundary[i] = 0;
}

static void copy_message(obuf* out, const char* filename)
{
  int fd = open_file("mess", filename);
  char buf[4096];
  if(fd == -1)
    die1sys(111, "Could not open message file");
  if(opt_mime)
    obuf_putf(out, mime_message_seperator, mime_boundary);
  else
    obuf_puts(out, message_seperator);
  if(opt_msgbytes < 0)
    for(;;) {
      ssize_t rd = read(fd, buf, sizeof buf);
      if(rd <= 0)
	break;
      if (!obuf_write(out, buf, rd))
	break;
    }
  else {
    ssize_t total = 0;
    while(total < opt_msgbytes) {
      ssize_t tord = ((unsigned)opt_msgbytes < sizeof buf) ?
	opt_msgbytes : (ssize_t)sizeof buf;
      ssize_t rd = read(fd, buf, tord);
      if(rd <= 0)
	break;
      if (!obuf_write(out, buf, rd))
	break;
      total += rd;
    }
    if(total >= opt_msgbytes)
      obuf_puts(out, "[...]");
  }
  close(fd);
  if(opt_mime)
    obuf_putf(out, mime_message_end, mime_boundary);
  obuf_putc(out, '\n');
}

void load_bounce_body(const char* filename)
{
  int in;
  struct stat sbuf;
  char* buf;
  if ((in = open(filename, O_RDONLY)) == -1)
    die3sys(111, "Could not open '", filename, "'");
  if (fstat(in, &sbuf) == -1)
    die3sys(111, "Could not stat '", filename, "'");
  if ((buf = malloc(sbuf.st_size)) == 0)
    die1(111, "Could not allocate bytes for loading bounce response file.");
  if (read(in, buf, sbuf.st_size) != sbuf.st_size)
    die1(111, "Could not read the bounce response file.");
  close(in);
  bounce_body = buf;
}
  
static void time2str(time_t secs, char* buf)
{
  const char* name;
  if(secs >= 24*60*60)
    name = " days", secs /= 24*60*60;
  else if(secs >= 60*60)
    name = " hours", secs /= 60*60;
  else if(secs >= 60)
    name = " minutes", secs /= 60;
  else
    name = " seconds";
  strcpy(buf, utoa(secs));
  strcat(buf, name);
}

void make_bounce_body(int fd, const char* sender, const char* filename,
		      const char* remotes, const char* locals, time_t age)
{
  const char* ptr;
  char time1[128];
  char time2[128];
  obuf out;

  obuf_init(&out, fd, 0, 0, 0);
  time2str(age, time1);
  time2str(queuelifetime, time2);

  obuf_putf(&out, bounce_header, me, sender);
  if(opt_mime) {
    make_mime_boundary();
    obuf_putf(&out, mime_bounce_header, mime_boundary, mime_boundary);
  }
  obuf_puts(&out, "\n");
  obuf_putf(&out, bounce_body, me, time1, time2);
  obuf_puts(&out, recipients_prefix);
  for(ptr = locals; *ptr; ptr += strlen(ptr)+1)
    if(*ptr == 'T')
      obuf_putf(&out, "{\t}s{\n}", ptr+1);
  for(ptr = remotes; *ptr; ptr += strlen(ptr)+1)
    if(*ptr == 'T')
      obuf_putf(&out, "{\t}s{\n}", ptr+1);
  if(opt_msgbytes)
    copy_message(&out, filename);
  if (!obuf_flush(&out))
    die1sys(111, "Could not write bounce message");
}


qmail-notify.c
/* qmail-notify -- Delayed delivery notification for qmail
 * Copyright (C) 2013  Bruce Guenter <bruceg@em.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
#include <sys/types.h>
#include <ctype.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>

#include <bglibs/sysdeps.h>
#include <bglibs/systime.h>
#include <bglibs/cdb.h>
#include <bglibs/cli.h>
#include <bglibs/dict.h>
#include <bglibs/iobuf.h>
#include <bglibs/misc.h>
#include <bglibs/msg.h>
#include <bglibs/wrap.h>
#include <bglibs/str.h>

#include "conf_qmail.h"
#include "qmail-notify-cli.h"
#include "qmail-notify.h"

const char program[] = "qmail-notify";

static const time_t default_age = 4*60*60;

static time_t now;
static time_t lastrun;
static pid_t inject_pid;
const char* me = 0;
time_t queuelifetime = 0;
static const char* extra_rcpt = 0;

static const char* qmail_home;
static const time_t* opt_ages = 0;
static unsigned opt_age_count = 0;

static dict rcpthosts;
static struct cdb morercpthosts;
static int morercpthosts_fd;
static str strbuf;

void oom(void) { die1(111, "Out of memory"); }

void parse_age(const char* s, const cli_option* o)
{
  char* end;
  time_t* n;
  if ((n = realloc((char*)opt_ages, (opt_age_count+1) * sizeof *n)) == 0)
    oom();
  n[opt_age_count] = strtoul(s, &end, 0);
  if (*end != 0) die2(111, "Specified age is not a number: ", s);
  ++opt_age_count;
  opt_ages = n;
  (void)o;
}

unsigned count_undone(const char* list)
{
  unsigned undone = 0;
  for(;;) {
    if(*list == 0)
      return undone;
    if(*list == 'T')
      ++undone;
    list += strlen(list) + 1;
  }
}

int check_rcpt(const char* sender)
{
  char* domain;
  
  if(!opt_checkrcpt)
    return 1;

  domain = strchr(sender, '@');
  if(!domain)
    return 0;

  for(++domain; domain; domain = strchr(domain+1, '.')) {
    if (!str_copys(&strbuf, domain)) oom();
    if (dict_get(&rcpthosts, &strbuf)) return 1;
    if (morercpthosts_fd != -1 &&
	cdb_find(&morercpthosts, strbuf.s, strbuf.len) != 0)
      return 1;
  }
  return 0;
}

int fork_inject(const char* sender)
{
  int p[2];

  debug6(1, "forking ", opt_qmail_inject, " -f '' -a -- '", sender, "' ", extra_rcpt);

  if(opt_nosend) {
    inject_pid = 0;
    return dup(1);
  }
  
  if (pipe(p) != 0)
    die1sys(111, "Could not create pipe");
  inject_pid = fork();
  switch(inject_pid) {
  case -1:
    die1sys(111, "Could not fork");
  case 0:
    wrap_chdir(qmail_home);
    close(p[1]);
    close(0);
    dup2(p[0], 0);
    close(p[0]);
    execl(opt_qmail_inject, opt_qmail_inject, "-f", "", "-a", "--", sender, extra_rcpt, NULL);
    die1sys(111, "Exec of qmail-inject failed");
  default:
    close(p[0]);
    return p[1];
  }
}

void wait_inject(void)
{
  int status;
  if(inject_pid) {
    if(waitpid(inject_pid, &status, WUNTRACED) == -1)
      die1sys(111, "Could not wait for qmail-inject to exit");
    if(!WIFEXITED(status))
      die1(111, "qmail-inject crashed");
    if(WEXITSTATUS(status))
      die1(111, "qmail-inject exited with an error");
  }
}

void send_bounce(const char* sender, const char* filename,
		 const char* remotes, const char* locals, time_t age)
{
  int fd = fork_inject(sender);
  make_bounce_body(fd, sender, filename, remotes, locals, age);
  if (close(fd) != 0)
    die1sys(111, "Writing to qmail-inject failed");
  wait_inject();
}

void make_bounce(const char* sender, const char* filename, time_t age)
{
  char* remotes = read_file("remote", filename);
  char* locals = read_file("local", filename);
  unsigned undone = count_undone(remotes) + count_undone(locals);
  debug6(1, "filename=", filename, " sender='", sender, "' undone=", utoa(undone));
  
  if(undone)
    send_bounce(sender, filename, remotes, locals, age);
  free(locals);
  free(remotes);
}

void scan_info(const char* filename)
{
  struct stat statbuf;
  char infoname[100];
  time_t expiry;
  int fd;
  unsigned i;
  strcpy(infoname, "info/");
  strcpy(infoname+5, filename);
  if((fd = open(infoname, O_RDONLY)) == -1 ||
     fstat(fd, &statbuf) == -1) {
    /* This could race with qmail-clean, ignore missing files. */
    if (errno == ENOENT)
      return;
    die3sys(111, "Can't open or stat info file '", infoname, "'");
  }
  /* Handle the file only if it's expiry time (creation time + opt_age)
     is before now and after the last run */
  for (i = 0; i < opt_age_count; ++i) {
    expiry = statbuf.st_mtime + opt_ages[i];
    debug4(1, "filename=", filename, " expiry=", utoa(expiry));
    if(expiry > now)
      debug1(1, "ignoring, has not yet expired");
    else if(expiry <= lastrun) {
      debug1(1, "ignoring, was previously expired");
      break;
    }
    else {
      /* Load the sender address from the info file */
      char* sender = malloc(statbuf.st_size);
      if (read(fd, sender, statbuf.st_size) != statbuf.st_size)
        die3sys(111, "Reading from info file '", infoname, "' failed");
      if(check_rcpt(sender+1))
	make_bounce(sender+1, filename, opt_ages[i]);
      else
	debug2(1, "ignoring, sender was not in rcpthosts: ", sender+1);
      free(sender);
      break;
    }
  }
  close(fd);
}

void scan_dir(const char* dirnum)
{
  DIR* dir;
  direntry* entry;
  char buf1[100];
  strcpy(buf1, "info/");
  strcpy(buf1+5, dirnum);
  if((dir = opendir(buf1)) == 0)
    die1sys(111, "Can't open queue directory");
  while((entry = readdir(dir)) != 0) {
    if(entry->d_name[0] != '.') {
      char filename[100];
      strcpy(filename, dirnum);
      strcat(filename, "/");
      strcat(filename, entry->d_name);
      scan_info(filename);
    }
  }
  closedir(dir);
}

void scan_queue(void)
{
  DIR* dir;
  direntry* entry;
  wrap_chdir(qmail_home);
  wrap_chdir("queue");
  if((dir = opendir("info")) == 0)
    die1sys(111, "Can't open queue directory");
  while((entry = readdir(dir)) != 0)
    if(entry->d_name[0] != '.')
      scan_dir(entry->d_name);
  closedir(dir);
}

static void load_rcpthosts(void)
{
  char* rh = read_file(0, "rcpthosts");
  if (!dict_init(&rcpthosts)) oom();
  if (rh) {
    const char* curr = rh;
    while (*curr) {
      const char* end;
      const char* next;
      if ((end = strchr(curr, '\n')) != 0)
	next = end + 1;
      else
	next = end = curr + strlen(curr);
      if (*curr != '#') {
	if (!str_copyb(&strbuf, curr, end-curr)) oom();
	if (!dict_add(&rcpthosts, &strbuf, 0)) oom();
      }
      curr = next;
    }
    free(rh);
  }
  if ((morercpthosts_fd = open_file(0, "morercpthosts.cdb")) != -1)
    cdb_init(&morercpthosts, morercpthosts_fd);
}

void load_config(void)
{
  unsigned i;

  wrap_chdir(qmail_home);
  wrap_chdir("control");
  
  me = read_line("me");
  if(!*me)
    die1(111, "Could not read control/me");
  queuelifetime = read_int("queuelifetime", 604800);
  now = time(0);
  lastrun = read_int(opt_run_file, 0);
  
  if(extra_rcpt_name && *extra_rcpt_name) {
    if(strchr(extra_rcpt_name, '@'))
      extra_rcpt = extra_rcpt_name;
    else {
      char* er = malloc(strlen(extra_rcpt_name)+strlen(me)+2);
      strcpy(er, extra_rcpt_name);
      strcat(er, "@");
      strcat(er, me);
      extra_rcpt = er;
    }
  }
  else
    extra_rcpt = 0;

  if (opt_checkrcpt) load_rcpthosts();

  debug3(1, "me='", me, "'");
  debug2(1, "queuelifetime=", utoa(queuelifetime));
  debug3(1, "extra_rcpt='", extra_rcpt, "'");
  debug2(1, "now=", utoa(now));
  debug2(1, "lastrun=", utoa(lastrun));
  for (i = 0; i < opt_age_count; ++i)
    debug2(1, "opt_ages[]=", utoa(opt_ages[i]));
}

void touch_run_file(void)
{
  obuf out;
  if (!obuf_open(&out, opt_run_file, OBUF_CREATE|OBUF_TRUNCATE, 0666, 0) ||
      !obuf_putu(&out, now) ||
      !obuf_close(&out))
    die1sys(111, "Could not update run file");
}

static int cmp_age(const void* aptr, const void* bptr)
{
  time_t a = *(time_t*)aptr;
  time_t b = *(time_t*)bptr;
  return b - a;
}

int cli_main(int argc, char* argv[])
{
  if (opt_debug)
    msg_debug_bits |= 0xff;
  if ((qmail_home = getenv("QMAILHOME")) == NULL)
    qmail_home = conf_qmail;
  if (!opt_ages) {
    opt_ages = &default_age;
    opt_age_count = 1;
  }
  else
    /* Sort the ages into descending order */
    qsort((void*)opt_ages, opt_age_count, sizeof *opt_ages, cmp_age);
  if (opt_bounce_filename != 0) load_bounce_body(opt_bounce_filename);
  load_config();
  scan_queue();
  if(!opt_nosend)
    touch_run_file();
  return 0;
  (void)argc;
  (void)argv;
}
